### ISSUE TITLE
Add sftp `connector`

### ISSUE THREAD

**Issue by fvankrieken** [NYCPlanning/data-engineering#1270]
## The problems

- We need to distribute PLUTO major releases to an SFTP server
- We need to ingest source data from SFTP servers
- We want to showcase the potential for various destination types with `dcpy.lifecycle`/metadata

## Acceptance Criteria
- [x] the SFTP connector can push, pull, and list files form SFTP servers (don't push to DOF's ginger server)
- [x] easy to run only unit tests or only integration tests

## Not In Scope

- Deciding exactly what "integration" means. New tests here will require dependencies to exist a server) unlike unit tests.
- Refactoring existing tests that require external dependencies 

### Models

In `dcpy/models/lifecycle/ingest.py` define a model for `SFTPServer`

### Connectors / Lifecycle

In `dcpy/connects/ftp.py` add functionality to `FTPConnector` stub to push and pull from an SFTP server.

[`products/pluto/pluto_build/bash/ssh.sh`](https://github.com/NYCPlanning/data-engineering/blob/main/products/pluto/pluto_build/bash/ssh.sh) is the example of how we currently pull from an SFTP server.

All servers we currently access in production require a private key, so that is the primary profile/credential configuration we need to handle. We currently:
- use 1password to store keys
- load them as environment variables
- save them as a file like `/.ssh/ginger`
- use them in invocations of the `sftp` CLI

The implementation of `FTPConnector` will align with this handling of credentials and use the python package [`pysftp`](https://pysftp.readthedocs.io/)

### Tests

Add tests `dcpy/test/connectors/` to confirm the push and pull of files (using similar patterns as other tests which confirm the existence of files)

Add new/use existing metadata and ingest templates to test the models and connector (e.g. `garlic_sftp` in multuple `dcpy/test/` locations)

Ensure `DistributionSFTPConnector` is as tested as `SocrataPublishConnector`

Use a dockerized SFTP server for tests to interact with instead of mocked responses.

### Nice to have

Run tests which require an SFTP server during PR tests in `test_helper.yml` via docker compose. If we can run relevant tests in a dev container locally, that may be good enough for now.

Use safeguards like [pytest-socket](https://github.com/miketheman/pytest-socket) to prevent undesired server calls.

Handling different types of server profile/credential configuration (e.g. username + password without a key).

---

There are a few needed use cases for this. 
1. distributing pluto to ginger server - see #1226
2. ingesting pluto inputs from same server - see #1261 

We'll also need to add ginger credentials to onepass if they're not there already, and configure a known host for ginger as well

There's some code in a few private repos that might be useful to use...

Tagging @alexrichey for visibility - didn't see an issue yet and wanted this tracked as a blocker for #1261

**Comment by damonmcc** [comment]
## scope (expected "files changed")
- add a connector
- to test, mock a server via docker compose

## notes

vocab: going with sftp over ftp

**Comment by alexrichey** [comment]
@damonmcc This is a good start. I think it'd be good to get a little more details about your intentions for a few different things:
1. What (if any) library are you thinking of using for sftp?
2. How are we going to accommodate different profiles/credentials for different SFTP servers? Related: how are we going to dispatch differently based on the destination in the metadata? 
3. Related... when would we want to pull credentials for, say, Ginger from 1pw? It would have to be somewhat dynamic, unless we want to have very specific Ginger distribution jobs.
4. We'd be adding a fairly new type of test (🎉!) but how are we going to separate it from the other tests? (To your point, figuring *everything* out for testing vocab is out of scope, but this does force us to make at least one choice here)

**Comment by alexrichey** [comment]
Thanks, this is excellent. cc @fvankrieken and @sf-dcp if y'all have any input

> If we can run relevant tests in a dev container locally, that may be good enough for now

This is good callout, and I'm facing a similar situation with the QA data inputs work. I think having something that can work locally with some futzing (and maybe some minimal instructions for how to run things) is a great starting point.

**Comment by fvankrieken** [comment]
Yeah - this came up a little bit in our pairing @alexrichey but for now I think just a docker-compose to set up the needed dependencies (outside of the context of any dev container) seems like it would be the simplest, one less docker container than using a dev image plus we get to keep the dev environments that we seem to be happy with.

**Comment by damonmcc** [comment]
new code and tests to pull files working locally in dev container. tried docker container outside of it but was hitting similar "connection refused" issues as in CI

**Comment by damonmcc** [comment]
thankfully `pystfp.Connection.put()` does return something ([docs](https://pysftp.readthedocs.io/en/release_0.2.9/pysftp.html#pysftp.Connection.put)). will see if the returned `SFTPAttributes` are useful for writing a good test of pushing a file to a serving

**Comment by sf-dcp** [comment]
<img width="950" height="466" alt="Image" src="https://github.com/user-attachments/assets/d820c3a3-26df-4ad9-8a4c-fae985a90696" />

Example of using the connector for distributing PLUTO to ginger

**Comment by sf-dcp** [comment]
Idea for sftp setup in ingest/distribution:

- Have 1 profile per sftp server (e.g. "ginger" profile for DOF) in 1Password with same variable names across profiles. The variables per sftp server are: `hostname`, `username`, `private_key` (actual file), and `known_hosts` (actual file). 
- In ingest/distribution template, specify sftp profile name (e.g. "ginger") and a key (file to pull from or distribute to). The profile name should correspond to what we have in 1Password. 
  
  Example:
  ```yml
  ingestion:
    source:
      type: sftp
      profile: ginger 
      key: remote_server/my_file.csv
      format: csv
  ```
- The actual setup code then will construct and pull all necessary variables from 1Password into env variables on local machine like this:
   `f"op://Data Engineering/{sftp_profile}/username"`.
  For ginger, it would be `f"op://Data Engineering/ginger/username"`)

This is a pretty lightweight setup not requiring custom connectors. We would just need to have one format for sftp profiles in 1Password.

**Comment by fvankrieken** [comment]
sftp put into operation in ingest in #1937
